fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java

```diff
@@ -382,4 +408,96 @@ public int compare(DelinquencyRange o1, DelinquencyRange o2) {
         return ranges;
     }
 
+    private void applyDelinquencyDetailsForLoanInstallments(final Loan loan, final DelinquencyBucket delinquencyBucket,
+            final Map<Long, CollectionData> installmentsCollectionData) {
+        for (Map.Entry<Long, CollectionData> installmentCollectionData : installmentsCollectionData.entrySet()) {
+            Optional<LoanRepaymentScheduleInstallment> installment = loan.getRepaymentScheduleInstallments().stream()
```

Comment:
This can be optimised a little bit:
Iterate through the installments of the loan and get by id from the installmentCollectionData the entry. If there is none, go next.

This way you only iterate through on the installments only once.

Reply:
done